File: Easy/1592. Rearrange Spaces Between Words/1592. Rearrange Spaces Between Words.py
class Solution:
    def reorderSpaces(self, text: str) -> str:
        text = " " + text + " "
        res = list(text)
        ans = []

        cnt = 0
        words = []
        i = 0

        while i < len(text) - 1:
            if res[i] == " " and i != 0:
                cnt += 1

            if res[i + 1] != " ":
                ft = i + 1
                i += 1
                while i < len(text) and res[i] != " ":
                    i += 1
                val = text[ft:i]
                words.append(val)
            else:
                i += 1

        sp = 0
        if len(words) - 1 != 0 and len(text) - 1 != 0:
            sp = cnt // (len(words) - 1)

        for word in words:
            ans.append(word)
            ans.append(" " * sp)

        rem = 0
        if len(words) - 1 != 0:
            rem = cnt % (len(words) - 1)
        if len(words) == 1:
            rem = cnt

        res1 = "".join(ans).rstrip()
        res1 += " " * rem

        return res1

File: Easy/1592. Rearrange Spaces Between Words/1592. Rearrange Spaces Between Words.rs
impl Solution {
    pub fn reorder_spaces(text: String) -> String {
        let words: Vec<&str> = text.trim().split_whitespace().collect();
        let word_count = words.len();
        let space_count = text.chars().filter(|&c| c == ' ').count();

        if word_count == 1 {
            return format!("{}{}", words[0], " ".repeat(space_count));
        }

        let spaces_between = space_count / (word_count - 1);
        let extra_spaces = space_count % (word_count - 1);
        let spaces_between_str = " ".repeat(spaces_between);

        let mut result = String::new();
        for i in 0..word_count {
            result.push_str(words[i]);
            if i < word_count - 1 {
                result.push_str(&spaces_between_str);
            }
        }

        result.push_str(&" ".repeat(extra_spaces));

        result
    }
}
